### PR TITLE
Implement Vsock in ACRN device model

### PR DESCRIPTION
**Pull request by liudlong**
1.  [ACRN: DM: Expose the vhost kernel ioctrl API]
2.  [ACRN: DM: Implement the vhost vsock]
3.  [ACRN: DM: Fix the virtio negotiated features bug.]

### CODE REVIEW

**Comment by acrnsi** [comment]
Can one of the admins verify this patch?

**Review by ywan170 — COMMENTED** [review]
@liudlong the commit message needs to be updated. Please check the comments from mailing list carefully.

**Comment by liudlong** [comment]
> @liudlong the commit message needs to be updated. Please check the comments from mailing list carefully.

@ywan170 Update the commit message and force update the PR.

**Review by ywan170 — APPROVED** [review]

**Comment by GGGZZZRRR** [comment]
@liudlong please check make doc error
+ make doc
make -C /home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/doc html BUILDDIR=/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/build/doc
make[1]: Entering directory '/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/doc'
making HTML content
doc build tool versions found on your system...

  breathe                   version: 4.31.0
  sphinx                    version: 3.5.4
  docutils                  version: 0.16
  sphinx-rtd-theme          version: 1.0.0
  sphinx-tabs               version: 1.3.0
  doxygen                   version: 1.8.17


New errors/warnings found, please fix them:
==============================================

/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:147: warning: end of file while inside a group
/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:134: warning: argument 'The' of command @param is not found in the argument list of vhost_kernel_ioctl(struct vhost_dev *vdev, unsigned long int request, void *arg)
/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:134: warning: argument 'The' of command @param is not found in the argument list of vhost_kernel_ioctl(struct vhost_dev *vdev, unsigned long int request, void *arg)
/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:145: warning: The following parameters of vhost_kernel_ioctl(struct vhost_dev *vdev, unsigned long int request, void *arg) are not documented:
  parameter 'request'
  parameter 'arg'
/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:134: warning: argument 'The' of command @param is not found in the argument list of vhost_kernel_ioctl(struct vhost_dev *vdev, unsigned long int request, void *arg)
/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:134: warning: argument 'The' of command @param is not found in the argument list of vhost_kernel_ioctl(struct vhost_dev *vdev, unsigned long int request, void *arg)
/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:145: warning: The following parameters of vhost_kernel_ioctl(struct vhost_dev *vdev, unsigned long int request, void *arg) are not documented:
  parameter 'request'
  parameter 'arg'
/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:134: warning: argument 'The' of command @param is not found in the argument list of vhost_kernel_ioctl(struct vhost_dev *vdev, unsigned long int request, void *arg)
/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:134: warning: argument 'The' of command @param is not found in the argument list of vhost_kernel_ioctl(struct vhost_dev *vdev, unsigned long int request, void *arg)
/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:145: warning: The following parameters of vhost_kernel_ioctl(struct vhost_dev *vdev, unsigned long int request, void *arg) are not documented:
  parameter 'request'
  parameter 'arg'
/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:134: warning: argument 'The' of command @param is not found in the argument list of vhost_kernel_ioctl(struct vhost_dev *vdev, unsigned long int request, void *arg)
/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:134: warning: argument 'The' of command @param is not found in the argument list of vhost_kernel_ioctl(struct vhost_dev *vdev, unsigned long int request, void *arg)
/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:145: warning: The following parameters of vhost_kernel_ioctl(struct vhost_dev *vdev, unsigned long int request, void *arg) are not documented:
  parameter 'request'
  parameter 'arg'
/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:134: warning: argument 'The' of command @param is not found in the argument list of vhost_kernel_ioctl(struct vhost_dev *vdev, unsigned long int request, void *arg)
/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:134: warning: argument 'The' of command @param is not found in the argument list of vhost_kernel_ioctl(struct vhost_dev *vdev, unsigned long int request, void *arg)
/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:145: warning: The following parameters of vhost_kernel_ioctl(struct vhost_dev *vdev, unsigned long int request, void *arg) are not documented:
  parameter 'request'
  parameter 'arg'
/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:134: warning: argument 'The' of command @param is not found in the argument list of vhost_kernel_ioctl(struct vhost_dev *vdev, unsigned long int request, void *arg)
/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:134: warning: argument 'The' of command @param is not found in the argument list of vhost_kernel_ioctl(struct vhost_dev *vdev, unsigned long int request, void *arg)
/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/devicemodel/include/vhost.h:145: warning: The following parameters of vhost_kernel_ioctl(struct vhost_dev *vdev, unsigned long int request, void *arg) are not documented:
  parameter 'request'
  parameter 'arg'

make[1]: *** [Makefile:64: html] Error 1
make[1]: Leaving directory '/home/acrn/workspace/acrn-hypervisor-build-doc_771/acrn-hypervisor/doc'
make: *** [Makefile:156: doc] Error 2

File: devicemodel/hw/pci/virtio/vhost_vsock.c [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+		pr_err(("vhost init out of memory!\n"));
+		goto fail;
+	}
+
+	/* pre-init before calling vhost_dev_init */
+	vhost_vsock->vdev.nvqs = 2;
+	vhost_vsock->vdev.vqs = vhost_vsock->vqs;
+	vhost_vsock->vhost_fd = open("/dev/vhost-vsock", O_RDWR);;
+	if (vhost_vsock->vhost_fd < 0) {
+		pr_err(("Open vhost-vsock fail, pls open vsock kernel config!\n"));
```
Comment (gvancuts) [inline devicemodel/hw/pci/virtio/vhost_vsock.c:184]: Why the double parentheses?

The message does not make a lot of sense. I guess you means something like:
```
Failed to open the vhost vsock file descriptor, please make sure your kernel configuration has the CONFIG_VHOST_VSOCK turned on.
```

File: devicemodel/hw/pci/virtio/vhost_vsock.c [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+virtio_vhost_vsock_init(struct vmctx *ctx, struct pci_vdev *dev, char *opts)
+{
+	struct virtio_vsock *vsock;
+	int rc, cid;
+	pthread_mutexattr_t attr;
+	char *devopts = NULL;
+	char *tmp = NULL;
+
+	if (opts == NULL) {
+		pr_err(("vsock: must have a valid guest_cid!\n"));
```
Comment (gvancuts) [inline devicemodel/hw/pci/virtio/vhost_vsock.c:247]: Remove the exclamation mark.

What is a `guest_cid`?

File: devicemodel/hw/pci/virtio/vhost_vsock.c [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+		return -1;
+	}
+	devopts = tmp = strdup(opts);
+	if (!devopts) {
+		pr_err(("vsock: The vsock parameter is NULL!\n"));
```
Comment (gvancuts) [inline devicemodel/hw/pci/virtio/vhost_vsock.c:252]: Ditto

File: devicemodel/hw/pci/virtio/vhost_vsock.c [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+		return -1;
+	}
+	if (!strncmp(tmp, "cid=", 4)) {
+		strsep(&tmp, "=");
+		dm_strtoi(tmp, NULL, 10, &cid);
+	}
+	free(devopts);
+
+	if (cid <= VMADDR_CID_HOST || cid >= U32_MAX) {
+		pr_err("vsock: guest_cid has to be 0x2~0xffffffff!\n");
```
Comment (gvancuts) [inline devicemodel/hw/pci/virtio/vhost_vsock.c:262]: Ditto

File: devicemodel/hw/pci/virtio/vhost_vsock.c [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+		return -1;
+	}
+
+	vsock = calloc(1, sizeof(struct virtio_vsock));
+	if (!vsock) {
+		pr_err("vosck: memory allocate failed!");
```
Comment (gvancuts) [inline devicemodel/hw/pci/virtio/vhost_vsock.c:268]: Ditto

File: devicemodel/hw/pci/virtio/vhost_vsock.c [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+		return -1;
+	}
+
+	vsock->config.guest_cid = cid;
+
+	rc = pthread_mutexattr_init(&attr);
+	if (rc)
+		pr_err("vsock: mutexattr init failed with erro %d\n", rc);
```
Comment (gvancuts) [inline devicemodel/hw/pci/virtio/vhost_vsock.c:276]: Ditto

Typo too: `erro` -> `error` (or `errno`)

File: devicemodel/hw/pci/virtio/vhost_vsock.c [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+	rc = pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE);
+	if (rc)
+		pr_err("vsock: mutexattr_settype failed with error %d\n", rc);
```
Comment (gvancuts) [inline devicemodel/hw/pci/virtio/vhost_vsock.c:279]: `mutexattr_settype` -> `pthread_mutexattr_settype`

File: devicemodel/include/vhost.h [hunk 8]
```diff
@@ -132,19 +132,15 @@ int vhost_dev_start(struct vhost_dev *vdev);
 int vhost_dev_stop(struct vhost_dev *vdev);
 
 /**
- * @brief set backend fd of vhost net.
+ * @brief vhost kernel dev ioctrl function.
```
Comment (gvancuts) [inline devicemodel/include/vhost.h:135]: `ioctrl` -> `ioctl`

You have the same typo in your commit message :-(

File: devicemodel/include/vhost_vsock.h [hunk 9]
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (C) 2018 Intel Corporation.
+ * Copyright (C) 2022 Intel Corporation.
```
Comment (gvancuts) [inline devicemodel/include/vhost_vsock.h:3]: These two lines should be combined into one if both were from Intel Corporation.

**Pull request merged**